internal/controller/pullrequest_controller.go

```diff
@@ -58,121 +58,101 @@ type PullRequestReconciler struct {
-	if pr.Status.State != "" && pr.Spec.State == pr.Status.State && pr.Status.ObservedGeneration == pr.Generation {
+	if pr.Status.State == pr.Spec.State && pr.Status.ObservedGeneration == pr.Generation {
 		logger.V(4).Info("Reconcile not needed")
 		return ctrl.Result{}, nil
 	}
 
-	// We want the PR to be open, but it's not open on the provider, we should create it.
-	if pr.Spec.State == promoterv1alpha1.PullRequestOpen && pr.Status.State != promoterv1alpha1.PullRequestOpen {
-		err = r.createPullRequest(ctx, &pr, pullRequestProvider)
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to create pull request: %w", err)
-		}
-	}
-
-	// We want to merge the PR, but it's not merged on the provider, we should merge it on the provicer and delete on k8s.
-	if pr.Spec.State == promoterv1alpha1.PullRequestMerged && pr.Status.State != promoterv1alpha1.PullRequestMerged {
-		err = r.mergePullRequest(ctx, &pr, pullRequestProvider)
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to merge pull request: %w", err)
-		}
-
-		err = r.Status().Update(ctx, &pr)
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("merged pull request in the SCM, but failed to update PullRequest status: %w", err)
-		}
-		// We requeue here because we want to delete the resource on the next reconcile, this does cause more API calls to the provider
-		// but is cleaner than deleting right away. Will revisit this.
-		return ctrl.Result{Requeue: true}, nil
-	}
-
-	// We want to close the PR, but it's not closed on the provider based on status, we should close it on provider and delete it from k8s.
-	if pr.Spec.State == promoterv1alpha1.PullRequestClosed && pr.Status.State != promoterv1alpha1.PullRequestClosed {
-		err := r.closePullRequest(ctx, &pr, pullRequestProvider)
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to close pull request: %w", err)
-		}
-
-		err = r.Status().Update(ctx, &pr)
-		if err != nil {
-			return ctrl.Result{}, fmt.Errorf("closed pull request, but failed to update PullRequest status: %w", err)
+	logger.Info("Reconciling PullRequest state", "desired", pr.Spec.State, "current", pr.Status.State)
+	if pr.Status.State != pr.Spec.State {
+		if pr.Spec.State == promoterv1alpha1.PullRequestOpen {
+			if pr.Status.ID == "" {
+				logger.Info("Creating PullRequest")
+				if err := r.createPullRequest(ctx, &pr, provider); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to create pull request: %w", err)
+				}
+			}
+		} else if pr.Spec.State == promoterv1alpha1.PullRequestMerged {
+			logger.Info("Merging PullRequest")
+			if err := r.mergePullRequest(ctx, &pr, provider); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to merge pull request: %w", err)
+			}
+			if err := r.Client.Delete(ctx, &pr); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to delete PullRequest: %w", err)
+			}
+			// No need to update the status here, as the pull request is deleted
+			return ctrl.Result{}, nil
```

Comment:
I think we can drop this now. The k8s call is cheap, and it'll make sure that on the next reconciliation we short-circuit before FindOpen.